controller_manager/src/controller_manager.cpp

```diff
@@ -53,61 +59,793 @@ ControllerManager::load_controller(
+  if (!has_parameter(param_name)) {
+    declare_parameter(param_name, rclcpp::ParameterValue());
+  }
+  if (!get_parameter(param_name, controller_type)) {
+    RCLCPP_ERROR(get_logger(), "'type' param not defined for %s", controller_name.c_str());
+    return nullptr;
+  }
+  return load_controller(controller_name, controller_type);
+}
+
+controller_interface::return_type ControllerManager::unload_controller(
+  const std::string & controller_name)
+{
+  // get reference to controller list
+  int free_controllers_list = (current_controllers_list_ + 1) % 2;
+  while (rclcpp::ok() && free_controllers_list == used_by_realtime_) {
+    if (!rclcpp::ok()) {
+      return controller_interface::return_type::ERROR;
+    }
+    std::this_thread::sleep_for(std::chrono::microseconds(200));
+  }
+  std::vector<ControllerSpec>
+  & from = controllers_lists_[current_controllers_list_],
+  & to = controllers_lists_[free_controllers_list];
+  to.clear();
+
+  // Transfers the running controllers over, skipping the one to be removed and the running ones.
+  bool removed = false;
+  for (const auto & controller : from) {
+    if (controller.info.name == controller_name) {
+      if (isControllerRunning(*controller.c)) {
+        to.clear();
+        RCLCPP_ERROR(
+          get_logger(),
+          "Could not unload controller with name '%s' because it is still running",
+          controller_name.c_str());
+        return controller_interface::return_type::ERROR;
+      }
+      RCLCPP_DEBUG(get_logger(), "Cleanup controller");
+      controller.c->get_lifecycle_node()->cleanup();
+      executor_->remove_node(controller.c->get_lifecycle_node()->get_node_base_interface());
+      removed = true;
+    } else {
+      to.push_back(controller);
+    }
+  }
+
+  // Fails if we could not remove the controllers
+  if (!removed) {
+    to.clear();
+    RCLCPP_ERROR(
+      get_logger(),
+      "Could not unload controller with name '%s' because no controller with this name exists",
+      controller_name.c_str());
+    return controller_interface::return_type::ERROR;
+  }
+
+
+  // Destroys the old controllers list when the realtime thread is finished with it.
+  RCLCPP_DEBUG(get_logger(), "Realtime switches over to new controller list");
+  int former_current_controllers_list_ = current_controllers_list_;
+  current_controllers_list_ = free_controllers_list;
+  while (rclcpp::ok() && used_by_realtime_ == former_current_controllers_list_) {
+    if (!rclcpp::ok()) {
+      return controller_interface::return_type::ERROR;
+    }
+    std::this_thread::sleep_for(std::chrono::microseconds(200));
+  }
+  RCLCPP_DEBUG(get_logger(), "Destruct controller");
+  from.clear();
+  RCLCPP_DEBUG(get_logger(), "Destruct controller finished");
+
+  RCLCPP_DEBUG(get_logger(), "Successfully unloaded controller '%s'", controller_name.c_str());
+  return controller_interface::return_type::SUCCESS;
+}
+
+std::vector<ControllerSpec> ControllerManager::get_loaded_controllers() const
```

Comment:
I'm not sure I understand the question, why not use this function on line 105?

First of all  I believe it's important to view there (line 105) the "magic" with the `controllers_lists_` indexes. 

Besides that, this creates a copy, and the code there needs to modify the original. We could return a non-const reference for that, but I believe viewing the indexes here is important.

We might rename the index variables and make it more evident what is going on, it took me a while to fully understand the purpose of it.

This list thing was a tricky one, I do not like it very much, but I wanted to minimize the changes done from ROS1 in this logic. 

It is code that has been thoroughly tested to be correct and real time safe, and once accepted noone is going to take another look at it, so I went with what I knew worked from ROS1 instead of trying to make something alternative.